File: backend/app/schemas/calendar.py
import uuid
from datetime import date as date_type
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class CalendarEventCreate(BaseModel):
    title: str
    description: str | None = None
    start_at: datetime
    end_at: datetime
    property_id: uuid.UUID | None = None


class CalendarEventUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    property_id: uuid.UUID | None = None


class CalendarEventInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    description: str | None
    start_at: datetime
    end_at: datetime
    property_id: uuid.UUID | None
    created_at: datetime


class CalendarEntry(BaseModel):
    kind: str
    title: str
    all_day: bool
    date: date_type | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    link: str | None = None
    event_id: uuid.UUID | None = None

File: frontend/src/lib/calendar.ts
import { apiFetch } from "@/lib/api";

export type CalendarKind = "lease_start" | "lease_end" | "rent_due" | "maintenance" | "event";

export interface CalendarEntry {
  kind: CalendarKind;
  title: string;
  all_day: boolean;
  date: string | null;
  start_at: string | null;
  end_at: string | null;
  link: string | null;
  event_id: string | null;
  description: string | null;
  property_id: string | null;
}

export interface CalendarEventInput {
  title: string;
  description?: string | null;
  start_at: string;
  end_at: string;
  property_id?: string | null;
}

export interface CalendarEventInfo extends CalendarEventInput {
  id: string;
  created_at: string;
}

export function listCalendar(start: string, end: string) {
  return apiFetch<CalendarEntry[]>(`/api/v1/calendar?start=${start}&end=${end}`);
}

export function createEvent(body: CalendarEventInput) {
  return apiFetch<CalendarEventInfo>("/api/v1/calendar/events", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function updateEvent(id: string, body: Partial<CalendarEventInput>) {
  return apiFetch<CalendarEventInfo>(`/api/v1/calendar/events/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export function deleteEvent(id: string) {
  return apiFetch<void>(`/api/v1/calendar/events/${id}`, { method: "DELETE" });
}
